File: council/providers/google.py
from collections.abc import Generator

from google import genai
from google.genai import types

from .base import Message, Provider


class GoogleProvider(Provider):
    def __init__(self, model_id: str, api_key: str):
        super().__init__(model_id, api_key)
        self.client = genai.Client(api_key=api_key)

    def _prepare(self, messages: list[Message]) -> tuple[str | None, list[types.Content]]:
        system_instruction = None
        contents = []
        for msg in messages:
            if msg.role == "system":
                system_instruction = msg.content
            elif msg.role == "user":
                contents.append(types.Content(role="user", parts=[types.Part(text=msg.content)]))
            elif msg.role == "assistant":
                contents.append(types.Content(role="model", parts=[types.Part(text=msg.content)]))
        return system_instruction, contents

    def send_message(self, messages: list[Message]) -> str:
        system_instruction, contents = self._prepare(messages)
        config = types.GenerateContentConfig(
            max_output_tokens=2048,
            system_instruction=system_instruction,
            thinking_config=types.ThinkingConfig(thinking_budget=1024),
        )
        response = self.client.models.generate_content(
            model=self.model_id,
            contents=contents,
            config=config,
        )
        return response.text

    def stream_message(self, messages: list[Message]) -> Generator[str, None, None]:
        system_instruction, contents = self._prepare(messages)
        config = types.GenerateContentConfig(
            max_output_tokens=2048,
            system_instruction=system_instruction,
            thinking_config=types.ThinkingConfig(thinking_budget=1024),
        )
        for chunk in self.client.models.generate_content_stream(
            model=self.model_id,
            contents=contents,
            config=config,
        ):
            if chunk.text:
                yield chunk.text

File: workers/src/providers/google.ts
import { GoogleGenAI } from '@google/genai';
import type { Message } from '../types.ts';

/**
 * Streams a response from Google Gemini via the unified @google/genai SDK.
 * System messages are separated and passed as systemInstruction.
 * Thinking budget mirrors the Python implementation (1024 tokens).
 */
export async function* streamGoogle(
  apiKey: string,
  modelId: string,
  messages: Message[],
): AsyncGenerator<string> {
  const ai = new GoogleGenAI({ apiKey });

  const systemMsg = messages.find((m) => m.role === 'system');
  const chatMessages = messages.filter((m) => m.role !== 'system');

  // Convert to Google Content format (user → "user", assistant → "model")
  const contents = chatMessages.map((m) => ({
    role: m.role === 'assistant' ? 'model' : 'user',
    parts: [{ text: m.content }],
  }));

  const response = ai.models.generateContentStream({
    model: modelId,
    contents,
    config: {
      ...(systemMsg ? { systemInstruction: systemMsg.content } : {}),
      maxOutputTokens: 2048,
      thinkingConfig: { thinkingBudget: 1024 },
    },
  });

  for await (const chunk of await response) {
    const text = chunk.text;
    if (text) yield text;
  }
}
